x/bank/keeper/msg_server.go

```diff
@@ -181,3 +181,39 @@ func (k msgServer) SetSendEnabled(goCtx context.Context, msg *types.MsgSetSendEn
 
 	return &types.MsgSetSendEnabledResponse{}, nil
 }
+
+func (k msgServer) Burn(goCtx context.Context, msg *types.MsgBurn) (*types.MsgBurnResponse, error) {
+	var (
+		from []byte
+		err  error
+	)
+
+	var coins sdk.Coins
+	for _, coin := range msg.Amount {
+		coins = append(coins, sdk.NewCoin(coin.Denom, coin.Amount))
+	}
+
+	if base, ok := k.Keeper.(BaseKeeper); ok {
+		from, err = base.ak.AddressCodec().StringToBytes(msg.FromAddress)
+		if err != nil {
+			return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid from address: %s", err)
+		}
+	} else {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("invalid keeper type: %T", k.Keeper)
+	}
```

Comment:
Given that this code is independent of coins adding and validity, I'd recommend we move it upfront then bring the coins related code below it where it is closer to home.